### DISCUSSION TITLE
Cmake changes: Now and into the future

### DISCUSSION THREAD

**Discussion by K20shores** [Unidata/netcdf-c#2845] [General]
Howdy!

As part of a pilot program between ACOM and Unidata with @WardF hosting me, I've been doing some work in the netCDF repository. I have some decent experience with cmake and the first (and maybe only thing since it's a short program) I am doing is trying to modernize cmake with the explicit goal of making consumption of netCDF by other projects easier, allowing `fetch_content` to be used, if it's not already available. 

Most of my knowledge has come from the book [Professional CMake: A Practical Guide](https://crascit.com/professional-cmake/) (which I *highly* recommend).

Much of what I'm doing will address some things in #2713.

Below is a list of PRs that I've already submitted and what they contained.

- #2835
  - Places the project version in the cmake `project` command which populates cmake internal variables and then uses them throughout the project
- #2838 
  - Part of enabling `fetch_content` is making it easy for cmake to manage dependencies, or at least find them. This PR pulled all of the dependencies out of the top level cmake and placed them into `cmake/depdendencies.cmake`
- #2842
  - I removed all of the functions and macros from the top level cmake file and placed them into `cmake/netcdf_functions_macros.cmake`
- #2847
  - moved the `netcdf` cmake target to the top level directory so that things like compile defintions, linked libraries, and include directories are set only on the cmake target and not globally
- #2890
  - shows that netcdf no longer adds hdf5 libraries to it's cmake and pkg config files that don't exist
- #2894
  - shows that netcdf can be included successfully with fetch content and adds an integration test for this, as well as documentation
- #2895
  - prefixes all cmake options with NETCDF_

Here a list of changes that I would to make over the remaining 2+ months that I have left. I'm listing them here so that the broader community can see what work is going on and comment about it so ensure that we arrive at a solution that works for everyone. I'm doing my best to keep the PRs small on purpose, though task 1. below will likely be the largest.

# What I'm working on right now

1. Part of the aforementioned book includes a recommend directory structure. 
```
:: tree . 
.
├── CMakeLists.txt
├── cmake
├── packaging
└── tests
```

as well as a recommended top level CMakeLists.txt setup, which you may recognize as something I'm working towards

```
# Preamble
cmake_minimum_required(VERSION 3.21)
project(MyProj)

# Project wide setup
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED YES)
set(CMAKE_CXX_EXTENSIONS NO)

# Externally provided content brought into the current scope.
include(dependencies.cmake)

# Main targets built by this project
add_subdirectory(src)

# Not typically needed if there is a parent project
if(PROJECT_IS_TOP_LEVEL)
    add_subdirectory(tests)
    add_subdirectory(packaging)
endif()
```

Explicitly, at present I am attempting to place all packaging into the `packaging` directory so that netCDF could be built and delivered entirely within cmake. At the same time, I've had to define the netCDF library as a top level target earlier on than before so that compile definitions are not applied globally.

This is a step in the direction of supporting #2843 since we would no longer need autotools to support libtools versioning.

# Future 

2. ensure that the install includes all necessary cmake things to generated a `FindNetDDF.cmake` file if it doesn't already and test/fix anything that doesn't work with `fetch_content`
4. Place all `option`s near the top of the file
5. move autotools definitions to an separate file
6. Rework the hdf5 package finding. If possible, place it into a FindHDF5.cmake file, or bump the cmake required version to 3.24 and bring in hdf5 with `fetch_content` if hdf5 supports this. If we can bump the version, then `fetch_content` will preferably use installed version of hdf5 before pulling it from git and compiling it from source
7. If it makes sense, place all `set` commands used for working variables shortly after the the `options`, but this 

Up until this point I've just done things, mostly because they've been small. But moving forward I would like to create issues for each thing I do, since, you know, that's the better and more appropriate way to do things.

@WardF, @DennisHeimbigner, @LecrisUT, @edhartnett, @ZedThree I would appreciate any feedback or direction you'd like to give.

Cmake conventions:

- lowercase function calls
- set compile definitions, include directories, only on targets
- prefix each option with `NETCDF_` (needs to be applied

**Comment by LecrisUT** [comment]
Overall 👍. Just some minor details which would probably already be addressed in the actual PRs

About the folder structure, I'm not sure how common `/packaging` is. I usually go with that content at the bottom of `src`, especially when it's a multi-project format. That would be useful if some files are to be copied as-is.

Also, instead of `if(PROJECT_IS_TOP_LEVEL)` go for individual options like `NETCDF_INSTALL` so that downstream has controll. I usually also include `NETCDF_TESTS` and `NETCDF_SHARED_LIBS`.

**Reply by ZedThree** [reply]
+1 for individual options that default to `PROJECT_IS_TOP_LEVEL`:

```cmake
option(NETCDF_TESTS "Build the tests" ${PROJECT_IS_TOP_LEVEL})
```

**Comment by LecrisUT** [comment]
Crucial point that you should add: namespacing of options and targets. Currently it is inconsistent. Among those it is also important to check all cache variables such that no `CMAKE_` ones are set as cache within the project (it will conflict with downstream).

**Comment by LecrisUT** [comment]
> If possible, place it into a FindHDF5.cmake file, or bump the cmake required version to 3.24 and bring in hdf5 with fetch_content if hdf5 supports this.

Probably you are already familiar, but don't use FetchContent within `Find<Package>.cmake`. Instead if you want `FetchContent(FIND_PACKAGE_ARGS)`, implement a small wrapper. [Here](https://github.com/LecrisUT/CMakeExtraUtils/blob/compat/FetchContent/cmake/compat/CEU_FetchContent.cmake)'s a simple implementation (note a few bugs, such as these should be `macro` instead of `function`)

**Reply by ZedThree** [reply]
I would also say it would hopefully be possible to entirely rely on CMake's built in `FindHDF5`. Essentially all of the logic for finding HDF5 currently could be as simple as:

```cmake
find_package(HDF5 ${HDF5_VERSION_REQUIRED} COMPONENTS C HL REQUIRED)
```

and then using the imported target `hdf5::hdf5`

**Reply by K20shores** [reply]
#2847 includes this change (thanks @ZedThree!) 

There are two related issues to hdf5 that I want to investigate next, listed here for posterity:

- https://github.com/Unidata/netcdf-c/issues/1444
- https://github.com/spack/spack/pull/42878
- and this: https://github.com/Unidata/netcdf-c/issues/2869

**Reply by K20shores** [reply]
Related, #2890

**Comment by ZedThree** [comment]
Thanks @K20shores, this looks really good! Maybe you could look at #2774, which I think is the minimal changes required to get `fetch_content` to work.

I've found [this CMake project template](https://github.com/cpp-best-practices/cmake_template) really useful, it has quite a few modules that can be dropped in, like coverage and sanitizers.
That template also moves all the `option`s to a separate file, which is quite nice.

Something that's not on your list that I think perhaps should be is removing all the global `add_definitions` and `include_directories`, and replacing them with `target_compile_definitions` and `target_link_libraries`

**Reply by K20shores** [reply]
#2894 shows that `fetch_content` indeed already works. I added some github actions to run an integration test to show that it works. Only the windows builds are failing, but that has nothing to do with netcdf, and everything to do with other libraries not being found or linking incorrectly. Hopefully those issue will be fixed soon

**Comment by DennisHeimbigner** [comment]
> prefix each option with NETCDF_ (needs to be applied)

I assume that this means that we specify options as "-DNETCDF_XXX".
This would cause a lot of  complaints from our users.

**Reply by LecrisUT** [reply]
Yes, but not having namespaced options means that the project cannot always be consumed via `FetchContent` which is the modern recommendation of how to link to projects. This issue is particularly important when there are chains of dependencies.

Unfortunately CMake does not have option deprecation, only target deprecation, and although such deprecation can be manually implemented, I am not sure it is worth it over proper communication.

In principle this change will primarily affect packagers, but they will find out about this breakage immediately as files will be missing. For the general users, they would be reading the documentation on how to build and what options are available anyway, so the renaming will not really affect them.

**Reply by K20shores** [reply]
#2895 addressed this

**Comment by K20shores** [comment]
Another possible thing to do is figure out how to remove the global `HAVE_CONFIG_H` https://github.com/Unidata/netcdf-c/pull/2847/files#r1469377349

**Reply by K20shores** [reply]
https://github.com/Unidata/netcdf-c/pull/2847#discussion_r1481721920

**Comment by K20shores** [comment]
Remove uses of glob: https://github.com/Unidata/netcdf-c/pull/2847/files#r1468249196

**Comment by K20shores** [comment]
Create an interface library to replace `TLL_LIBS` https://github.com/Unidata/netcdf-c/pull/2847#issuecomment-1961092585

**Comment by K20shores** [comment]
use build in modules to find the jpeg library: https://github.com/Unidata/netcdf-c/pull/2847/files#r1468245297
